### PR TITLE
Add transformed distribution

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
split from #4678  (#3622 ).
By using `TransformedDistribution` and `Bijector`, we can implement (normalizing) flow such as MAF and Glow easier.

### CODE REVIEW

File: chainer/distributions/transformed_distribution.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    def event_shape(self):
+        return self.base_distribution.event_shape
+
+    def cdf(self, x):
+        return self.base_distribution.cdf(self.bijector.inv(x))
+
+    def log_prob(self, x):
+        return self.base_distribution.log_prob(self.bijector.inv(x)) \
+            - self.bijector.logdet_jac(self.bijector.inv(x))
+
```
Comment (toslunar) [inline chainer/distributions/transformed_distribution.py]: Could you also implement `prob`?

File: chainer/distributions/transformed_distribution.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    @property
+    def event_shape(self):
+        return self.base_distribution.event_shape
+
+    def cdf(self, x):
+        return self.base_distribution.cdf(self.bijector.inv(x))
+
+    def log_prob(self, x):
+        return self.base_distribution.log_prob(self.bijector.inv(x)) \
```
Comment (toslunar) [inline chainer/distributions/transformed_distribution.py]: Do not repeat `self.bijector.inv(x)`.

File: tests/chainer_tests/distributions_tests/test_transformed_distribution.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@testing.parameterize(*testing.product({
+    'shape': [(3, 2), (1,)],
+    'is_variable': [True, False],
+    'sample_shape': [(3, 2), ()],
+}))
+@testing.fix_random()
+@testing.with_requires('scipy')
+class TestLogTransformedDistribution(testing.distribution_unittest):
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_transformed_distribution.py]: This class tests `Exp`-transformed `Normal` distribution, which is a log-normal distribution.

File: chainer/distributions/bijector.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            return old_y
+        y = self._forward(x)
+        self.cache_x_y = x, y
+        return y
+
+    def inv(self, y):
+        old_x, old_y = self.cache_x_y
+        if id(y) == id(old_y):
+            return old_x
+        x = self._inv(y)
```
Comment (toslunar) [inline chainer/distributions/bijector.py]: `y` can be a `ndarray` with the current implementation of `TransformedDistribution.cdf`, while `self._inv` requires `Variable`.

File: chainer/distributions/bijector.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    """Interface of Bijector.
+
+    `Bijector` is implementation of bijective (invertible) function that is
+    used by `TransformedDistribution`. The three method `_forward`, `_inv` and
+    `_logdet_jac` have to be defined in inhereted class.
+    """
+
+    def __init__(self):
+        self.cache_x_y = (None, None)
```
Comment (toslunar) [inline chainer/distributions/bijector.py]: This caching mechanism seems a trade-off between speed and memory.  Users may want to turn off this feature.
This issue can be addressed in another PR.

Comment (toslunar) [reply chainer/distributions/bijector.py]: Wait.  There's a more significant issue.

In the current implementation of `TransformedDistribution.cdf`, we cannot compute the following.
```
d = TransformedDistribution(...)
y = d.sample(...)
q = d.cdf(y)
g, = chainer.grad(outputs=(q,), inputs=(y,))
```

Comment (YoshikawaMasashi) [reply chainer/distributions/bijector.py]: Oh, no...
Caching does not retain computational graph between input and output. But in many cases, `inv`  calculation is heavy and computational graph between input and output is needless.
Is it good solution that we add option to using cache? or is there any good solution that we can calculate gradient and cache?

File: chainer/distributions/bijector.py (hunk 4 above)
Comment (YoshikawaMasashi) [reply chainer/distributions/bijector.py]: I fix docstring as like that `self._inv` requires `Variable` and `ndarray`.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by ganow** [comment]
This PR is very attractive to me, but the development seems to be stopped due to the following problem (https://github.com/chainer/chainer/pull/5085#discussion_r201650176).

> Wait. There's a more significant issue.
> 
> In the current implementation of `TransformedDistribution.cdf`, we cannot compute the following.
> 
> ```
> d = TransformedDistribution(...)
> y = d.sample(...)
> q = d.cdf(y)
> g, = chainer.grad(outputs=(q,), inputs=(y,))
> ```

Pytorch avoids this problem by setting the flag for caching, and I think we can use the same technique.
https://github.com/pytorch/pytorch/blob/6a297b86754324d850d1595947c5e5559b51acc2/torch/distributions/transforms.py#L51-L52

**Comment by toslunar** [comment]
It's fine to drop the cache feature.  IMO, unsafe feature should be added with very good documentation.

**Pull request closed without merging**